### ISSUE TITLE
deep populate

### ISSUE THREAD

**Issue by MattMcFarland** [scottwrobinson/camo#13]
Hey,

Spent the last 6 hours trying to figure out how to deep populate a comments model.  Nested models are not getting populated.  With mongoose we can use a plugin called deepPopulate or populateDeep, and it works.  But with Camo (I love camo btw) there really isnt a way to do it.  I tried really hard and nothing works.  Please let me know if you have any ideas..

**Comment by scottwrobinson** [comment]
Hi Matt,

Glad you like Camo! It is still in the early stages, so there are definitely some missing features, like support for nested data is partially missing. I say 'partial' because you _can_ save nested data via `EmbeddedDocument`, but you can't specify a nested schema directly in a `Document`.

So, as of v0.8.0, you **can** do this:

```javascript
class Comment extends EmbeddedDocument {
    constructor() {
        super();
        this.username = String;
        this.body = String;
    }
}

class Post extends Document {
    constructor() {
        super('post');
        this.title = String;
        this.body = String;
        this.comments = [Comment];
    }
}

var post = Post.create({
    title: 'Camo is Fun',
    body: 'This is my post...',
    comments: []
});

var comment1 = Comment.create({
    username: 'scott',
    body: 'What a great post...'
});

var comment2 = Comment.create({
    username: 'bill',
    body: 'This post sucks'
});

post.comments.push(comment1);
post.comments.push(comment2);

post.save().then(function() {
    // ...
});
```

But you **can't** do this (yet):

```javascript
class Post extends Document {
    constructor() {
        super('post');
        this.title = String;
        this.body = String;
        this.comments = [{
            username: String,
            body: String
        }];
    }
}
```

The main difference is where the `Comment` schema is declared.

Hopefully the `EmbeddedDocument` will work out for you for now. If you have any more questions it might help if I saw the schema you're trying to create and I'll give you hints on how to do it. Otherwise I'm working on better nested data support so hopefully your problems will be solved soon.

Scott

**Comment by MattMcFarland** [comment]
Hey, thanks for the reply!

Yeah I was thinking about using Embedded Document but really the Comment Class is extended off of a Document class.  Reason is there are other document classes that have similar functionality, and I have it set up to where you can add a comment to a comment.  

So in order for me to use embedded document, I would have to duplicate a lot of code.  Really trying to avoid that right now, in fact spent 6+ hours trying to haha.

I did look at the Embedded Doc class and the code used to create the embed document (looks like it replaces the ref with actual doc maybe, permanently??) is quite complicated so then I didnt want to copy that either.

I ended up trying to use Promises to populate the code, but no matter what I do, it seems like it either creates a ton of copies of comments, or creates a weird nested array.  It just wasn't working! 

I do thank you for your time and I am a huge fan of Camo.  IT is really been amazing to use, much better than Mongoose (syntactically) 

I'll try to look at my code again, but really want to avoid Embed if I can.  I'll also share if I can make any headway.  So far it's been very challenging (which is good too :)) - You can actually see my adventures with camo (and struggles) https://www.livecoding.tv/mmcfarland/

**Comment by MattMcFarland** [comment]
So anyway, Here's the code I've struggling with:

```javascript
 get DTO () {
    return new Promise((resolve, reject) => {
      var solutionPromises = this._values.solutions.map((sol) => {
        return Promise.all(sol._values.comments.map((com) => {
          var solution = Object.assign(sol.DTO, {
            comments: []
          });
          return new Promise((res, rej) => {
            Comment.loadOne({_id: com}).then((comment) => {
              solution.comments[solution.comments.length] = (Object.assign(comment, comment.DTO));
              res(solution)
            })
          })
        }))
      });

      Promise.all(solutionPromises).then((solutions) => {
        resolve({
          solutions,
          id: this.id,
          title: this.title,
          linkMeta: this.linkMeta,
          authorName: this.authorName,
          authorUrl: this.authorUrl,
          editorName: this.editorName,
          editorUrl: this.editorUrl,
          flags: this.flags,
          score: this.tallyVotes(),
          comments: this._values.comments
        }, this._values)
      });
    });

  }
```

That is basically my "Question" class - as I'm creating a Q&A website.   (but my questions are known as tutorial requests, and answers are known as tutorial solutions)

Since Questions and Answers can both have comments, it's really the question replies that get screwed (as they are a child of questions) 

So the code above is really my 20th attempt (or so) at getting the comments to appear.  Like I said, I do get them to appear, but i end up making lots of copies.  Like the example above somehow makes a nested array of "solutions" then the comments for the solutions somehow get copied - what I mean is there's only two comments and one is on one solution each, but somehow I end up with like 8 comments and they're all the 4th comment (maybe I am doing some pass-by-value mutation)

**Comment by scottwrobinson** [comment]
Hmm I'm not sure I'm following your code exactly, but I _think_ I get what you're trying to do. Are you just trying to load all comments in each Solution?

A couple things:
- A `Proxy` is used in the background of `Document` to access your data from `_values`. It's the thing that makes `this.data === this._values.data` work.
- I would advise against accessing data directly via `this._values`. If you want to get the values stored in `sol`, just use `this.sol` and the `Proxy` object will handle retrieving it for you. The `Proxy` doesn't do a whole lot right now, but it will in future versions.
- I would also advise against merging objects via `Object.assign` (although I'm not really experienced on how it works). I say this mostly because I don't know how it affects the `Document`'s use of `Proxy` during assignment/retrieval.
- As of v0.8.0, all 1-deep embedded _and_ referenced documents are already populated for you. So if you have something like `Solution.comments` then using `loadOne` will already populate the `comments` for you, but `Solution.comments[0].comment` would not be populated. You'll have to load 2-deep (and greater) documents yourself.

Also, the part where you're doing...

```javascript
var solution = Object.assign(sol.DTO, {
    comments: []
});
```

...I think you're reassigning `comments` to an empty array in `sol` for every comment you iterate over.

Do your `DTO` methods just populate referenced data? If so, why are you doing `Object.assign(comment, comment.DTO)` after already loading the comment using `loadOne`? To get the further nested comments?

I might be of more help if I could see the actual schema for `TutorialSolution` and `Comment`.

Some of the problems of loading nested data like this is because of Camo's deficiencies, so this will be a good use case to keep in mind during development. Thanks!

**Comment by MattMcFarland** [comment]
Hey Scott - I really appreciate your time analyzing my code.  You've made some good points regarding Object.assign, and I haven't really messed with Proxies so I can say that I am not sure either.  I can say that object.assign wasn't behaving like I was intending for it.  However one thing I was trying to do with DTO is send the data upstream without sending _schema or _values, but rather the DTO is like a normalized transfer object that is brief.  some of the data coming in from DTO is also retrieved via `getters`.  

One thing you said here:

```
You'll have to load 2-deep (and greater) documents yourself
```

That is exactly what I'm trying to do, and have been failing at it.  The 1-deep comments load just fine.

Here's the base class, it is extended from all other documents and why I dont want to use embedds:

```javascript
"use strict";
// Base Model
const
  _        = require('lodash'),
  Document = require('camo').Document,
  Utils    = require('../utils');

class Base extends Document {
  constructor(name) {
    super(name);

    this.schema({
      author:   { type: Object },
      editor:   { type: Object },
      title:    { type: String },
      content:  { type: String },

      /* Flagging */
      flaggers:         { type: [Object]},

      /* Voting */
      voters: { type: [Object] },

      /* Meta */
      created_at: { type: Date, default: Date.now },
      updated_at: { type: Date, default: Date.now },
      removed:    { type: Boolean, default: false}
    });

  }



  get flags () {
    return {
      spam: this.countFlags('spam'),
      offensive: this.countFlags('offensive'),
      vague: this.countFlags('vague'),
      duplicate: this.countFlags('duplicate')
    }
  }
  get authorName () {
    if (!this.author) {
      return '';
    }
    return this.author.fullName;
  }

  get editorName () {
    if (!this.editor) {
      return '';
    }
    return this.editor.fullName;
  }

  get authorUrl () {
    if (!this.author) {
      return '';
    }
    return 'users/' + Utils.Users.getId(this.author);
  }

  get editorUrl () {
    if (!this.editor) {
      return '';
    }
    return 'users/' + Utils.Users.getId(this.author);
  }

  countFlags (flagType) {
    return _.countBy(this.flaggers, {type: flagType}).true || 0;
  }

  addOrRemoveFlag (user, flagType) {
    return new Promise((resolve, reject) => {
      var flagIndex, flag, result;
      try {
        this.flaggers || (this.flaggers = []);

        flagIndex = _.findIndex(this.flaggers, {'uid': Utils.Users.getId(user), type: flagType});
        flag = flagIndex > -1 ? this.flaggers[flagIndex] : undefined;
        result = {
          flagType: flagType,
          isFlagged: null
        };

        if (flag) {
          this.flaggers.splice(flagIndex, 1);
          result.isFlagged = false;
        } else if (typeof flag === "undefined") {
          this.flaggers.push({
            uid: Utils.Users.getId(user),
            type: flagType
          });
          result.isFlagged = true;
        }
        // signal back the new flag
        user.save();
        this.save()
          .then(() => resolve(result))
          .catch((e) => reject(e));
      } catch (e) { reject(e) }
    });

  }

  createOrUpdateVote (user, direction) {
    return new Promise((resolve, reject) => {
      var voteIndex, newVote, oldVote;

      try {
        this.voters || (this.voters = []);
        voteIndex = _.findIndex(this.voters, {'uid': Utils.Users.getId(user)});

        if (voteIndex === -1) {
          newVote = direction === "up" ? 1 : -1;
        } else {
          oldVote = this.voters[voteIndex].vote;
          if (direction === "up") {
            newVote = oldVote === 1 ? 0 : 1;
          } else if (direction === "down") {
            newVote = oldVote === -1 ? 0 : -1;
          }
          this.voters.splice(voteIndex, 1);
        }

        this.voters.push({
          uid: Utils.Users.getId(user),
          vote: newVote
        });

        this.save()
          .then((i) => {
          resolve({
            "score": i.score,
            "userVote": newVote
          });
        }).catch((e) => reject(e));
      } catch (e) { reject(e) }
    });
  }

  tallyVotes () {
    var score = 0;
    if (this.voters && this.voters.length) {
      this.voters.forEach((voter) => {
        score += voter.vote;
      });
    }
    return score;
  }

  removeOrUndoRemove (editor) {
    return new Promise((resolve, reject) => {
      try {
        this.updated_at = Date.now();
        this.editor = editor;
        this.removed = !this.removed;
        this.save()
          .then((i) => resolve(i))
          .catch((e) => reject(e));
      } catch (e) { reject(e) }
    });
  }

}

module.exports = Base;
```

As you can see, no DTO here.  The DTO's have officially been cleaned up, and now the data I Send upstream is very clean but I am not getting further nested comments, I cannot figure out how without blowing evrything up haha.

```javascript
Comment DTO:
  get DTO () {
    return {
      type: 'Comment',
      id: this.id,
      authorName: this.authorName,
      authorUrl: this.authorUrl,
      editorName: this.editorName,
      editorUrl: this.editorUrl,
      message: this.message,
      flags: this.flags,
      score: this.tallyVotes()
    }
  }
```

TutorialRequest DTO (most complex)
```javascript
 get DTO () {

    return {
      type: "TutorialRequest",
      solutions: this.solutions.map((sol) => {
        return sol.DTO;
      }),
      id: this.id,
      title: this.title,
      linkMeta: this.linkMeta,
      authorName: this.authorName,
      authorUrl: this.authorUrl,
      editorName: this.editorName,
      editorUrl: this.editorUrl,
      flags: this.flags,
      score: this.tallyVotes(),
      comments: this.comments.map((com) => {
        return com.DTO;
      })
    };


  }
```

```javascript
  get DTO () {
    return {
      type: 'TutorialSolution',
      id: this.id,
      authorName: this.authorName,
      authorUrl: this.authorUrl,
      editorName: this.editorName,
      editorUrl: this.editorUrl,
      flags: this.flags,
      score: this.tallyVotes(),
      comments: this.comments.map((com) => {
        if (com.DTO) {
          return com.DTO;
        } else {
          return Comment.loadOne({_id: com})
            .then((m) => m.DTO)
            .catch((e) => Utils.Log.error(e));
        }
      })
    }
  }
```

So basically when I download a tutorial request, I get the array of solutions which also have arrays of comments themselves (2 levels deep) - this is where mongoose plugin `populateDeep` would come in handy.

I tried to populate deep myself, and I ran into a ton of issues.  So the new DTO you see here are now simplified to the point to where I am no longer trying to populate deep.

The plan is to try to figure out how to populate deep without having to break things.  Also I am concerned that you recommend I avoid using `_values` - I would prefer to just get the values instead of schema, so do I just use `this.data` instead?

Thanks again, hope this makes sense.

**Comment by scottwrobinson** [comment]
Hi Matt,

That's one of the fundamental problems with JavaScript being asynchronous, if a method makes an async call, then it has to return a promise, no way around it.

It would definitely be nice if Camo could handle deep population. I'm still working on a way to let you tell Camo which fields to populate (and how deep) and which to leave as references. Hopefully I'll have the feature out soon. I'll take a look at Mongoose's populateDeep plugin. Thanks for pointing that out!

I should have more time to look at your code later this week. I'll get back to you then. Also, thanks for your feedback on Camo! It really helps to have another set of eyes and to see how people are using it.

Scott

**Comment by MattMcFarland** [comment]
Hey man,

I finally got this working! The solution below is a bit ugly, so I plan on cleaning it up later.. For now, I'm leaving it as is seeing as how I spent over 10 hrs on this problem..

```javscript
exports.getById = (M, req, res, Comment) => {
  return M.loadOne({_id: ObjectID(req.params.id)})
    .then((m) => {
      var
        done = false,
        solutions = _.clone(m.DTO.solutions),
        commentsLoaded = false,
        loop = (solIndex) => {
          var
            index = solIndex || 0,
            sol = solutions[index],
            j = 0;

          if (sol && sol.comments && sol.comments.length) {
            if (!sol.processing) {
              j = 0;
              sol.processing = true;
              //console.log('solution [' + index + '] has', sol.comments.length, 'comment(s)');
              sol.xcomments = Promise.all(sol.comments.map((c, i) => {
                //console.log('get comment', c);
                return Comment.loadOne({_id: c});
              })).then((g) => {
                commentsLoaded = true;
                sol.comments = g.map((com) => {
                  return com.DTO;
                });

                //console.log(g.length, 'comments loaded');
                if (done) {
                  res.json(Object.assign(m.DTO, {solutions}));
                }
              })
            } else if (done) {
              //console.log('but we done');
            } else {
              //console.log('waiting...');
            }
          } else {
            commentsLoaded = true;
            //console.log('solution [' + index + '] has no comment(s)');
          }
          if (!commentsLoaded) {
            _.defer(loop, 500);
          } else {
            if (index >= solutions.length - 1) {
              done = true;
            } else {
              loop(index + 1);
            }
          }
        };
      if (solutions.length) {
        //Utils.Log.info('found ' + solutions.length + ' solutions');
        loop();
      } else {
        res.json(m.DTO);
      }
    })
    .catch((e) => Utils.Log.error(e));
};
```

**Comment by scottwrobinson** [comment]
Awesome! Glad you got it working. Hopefully I can add extend Camo so you won't have to go through all of this again. Thanks again for all the feedback and everything.

I'm going to close this issue for now, but let me know if you need anything else.

Scott
